### PR TITLE
Dynamic generic dictionary expansion feature

### PR DESCRIPTION
**Pull request by fadimounir**
TODO:
- [x] Update JITInterface GUID in corinfo.h after https://github.com/dotnet/runtime/pull/1378#issuecomment-580944654

These changes introduce dynamic size expansion for generic dictionary layouts when we run out of slots.
The original implementation allowed for an expansion, but using a linked list structure, which made it
impossible to use fast lookup slots once we're out of slots in the first bucket.

This new implementation allows for the usage of fast lookup slots always, for all generic lookups.

This also removes the constraint we had on R2R, where we disabled the usage of fast slots all-together.

**_NOTE:_**
This is a much simpler implementation than the original commit which got reverted due to failures related to collectible assemblies (https://github.com/dotnet/coreclr/pull/26262). The difference between these changes and the original ones is that dictionary expansion for types and methods is now done lazily, on-demand, whenever we attempt to read a value beyond the size of a given dictionary. This makes uses of the newly added JIT support from https://github.com/dotnet/runtime/pull/31957. The bug with collectible assemblies have been also fixed, and I added a test to cover various interesting aspects of the feature.

**_BONUS:_**
Part of these changes is some cleanup work to really simplify the R2R dictionary access stubs. These stubs are now much more compact with fewer instructions and better registry usage (many #ifdef UNIX_AMD64_ABI blocks removed from the amd64 stub)

cc @dotnet/crossgen-contrib

### CODE REVIEW

File: src/coreclr/src/inc/corinfo.h
Comment (jkotas) [inline src/coreclr/src/inc/corinfo.h]: This is needs JIT interface GUIID change

File: src/coreclr/src/inc/corinfo.h
Comment (jkotas) [inline src/coreclr/src/inc/corinfo.h]: You will need the same change in crossgen2 src\coreclr\src\tools\Common\JitInterface\CorInfoTypes.cs

Comment (jkotas) [inline src/coreclr/src/inc/corinfo.h]: Nit: The struct would have better packing if this field is before offsets

File: src/coreclr/src/jit/compiler.h
Comment (jkotas) [inline src/coreclr/src/jit/compiler.h]: Why is this not using `CORINFO_SKIPSIZECHECK` ?

File: src/coreclr/src/inc/corinfo.h
Comment (jkotas) [inline src/coreclr/src/inc/corinfo.h]: Nit: A better name may be `CORINFO_NO_SIZE_CHECK`. 

The names on JIT/EE tend to use NO prefix to describe that something should not happen, e.g `CORINFO_NO_THIS_TRANSFORM`, `CORJIT_FLAG_NO_INLINING`, ...

File: src/coreclr/src/vm/genmeth.cpp
Comment (jkotas) [inline src/coreclr/src/vm/genmeth.cpp]: This does not seem to be DACized correctly. 

You may want to move this logic into `class Dictionary`. There are other similar helpers already.

File: src/coreclr/src/vm/methodtable.cpp
Comment (jkotas) [inline src/coreclr/src/vm/methodtable.cpp]: Dtto

**Review by jkotas — CHANGES_REQUESTED** [review]

**Comment by jkotas** [comment]
Overall, I like this much better than the original version. Thanks!

File: src/coreclr/src/inc/corinfo.h
Comment (MichalStrehovsky) [reply src/coreclr/src/inc/corinfo.h]: Note the JIT team requested holding off checking in JitInterface changes: https://github.com/dotnet/runtime/pull/1378#issuecomment-580944654

Pull requests #1378, #2249, #31686 are currently waiting for the window to reopen again.

Cc @BruceForstall on whether the window is going to open soon.

File: src/coreclr/src/inc/corinfo.h
Comment (fadimounir) [reply src/coreclr/src/inc/corinfo.h]: Fixed

File: src/coreclr/src/inc/corinfo.h
Comment (fadimounir) [reply src/coreclr/src/inc/corinfo.h]: Renamed

File: src/coreclr/src/jit/compiler.h
Comment (fadimounir) [reply src/coreclr/src/jit/compiler.h]: All the changes related to this feature on the JIT side had the "expruntimelookup" naming, so I wanted to follow the same convention on the JIT side for consistency. Would it be better to change it to CORINFO_NO_SIZE_CHECK?

File: src/coreclr/src/vm/genmeth.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/genmeth.cpp]: But this is under !DACCESS_COMPILE. Shouldn't that be enough?

File: src/coreclr/src/inc/corinfo.h
Comment (fadimounir) [reply src/coreclr/src/inc/corinfo.h]: Thanks for the heads-up Michal

Comment (BruceForstall) [reply src/coreclr/src/inc/corinfo.h]: We're ready to take JIT-EE interface changes now (some internal processes dependent on the JIT-EE interface GUID are now stable).

Please wait until #2249 is merged, then update the JIT-EE interface GUID, then feel free to merge.

**Comment by fadimounir** [comment]
Couple of interesting benchmark numbers I tried:

- Roslyn compiler, compiling Roslyn source code: **1~2%** performance gain.
- NewtonSoft serialization benchmark (found something on the web): **4%** performance gain.

File: src/coreclr/src/jit/compiler.h
Comment (jkotas) [reply src/coreclr/src/jit/compiler.h]: Yes. The JIT should be using the constants from corinfo.h. It should not be duplicating them with different names.

File: src/coreclr/src/vm/class.cpp
Comment (jkotas) [inline src/coreclr/src/vm/class.cpp]: Do we still need this?

File: src/coreclr/src/vm/genericdict.h
Comment (jkotas) [inline src/coreclr/src/vm/genericdict.h]:
```suggestion
    // Current number of non-type-argument slots
```

File: src/coreclr/src/vm/genericdict.h
Comment (jkotas) [inline src/coreclr/src/vm/genericdict.h]:
```suggestion
    // Create an initial dictionary layout containing numSlots slots
```

File: src/coreclr/src/vm/generics.cpp
Comment (jkotas) [inline src/coreclr/src/vm/generics.cpp]: `cbInstAndDict` cannot be zero here.

Comment (jkotas) [reply src/coreclr/src/vm/generics.cpp]: And `pOldMT->GetClass()->GetDictionaryLayout()` is likely too complex for the compiler to CSE so it would be nice to fetch it just once.

File: src/coreclr/src/vm/genmeth.cpp
Comment (jkotas) [reply src/coreclr/src/vm/genmeth.cpp]: You are right - it is not necessary to DACize for correctness.

But it would be still nice to have it on Dictionary like all other similar methods; and use  `TADDR` instead of `ULONG_PTR` since that's the type used to generic dictionary slots.

File: src/coreclr/src/vm/i386/cgenx86.cpp
Comment (jkotas) [inline src/coreclr/src/vm/i386/cgenx86.cpp]: What does `JLE` stand for?

File: src/coreclr/src/vm/genericdict.cpp
Comment (jkotas) [inline src/coreclr/src/vm/genericdict.cpp]: Is this still needed with the simpler scheme?

File: src/coreclr/src/vm/genericdict.cpp
Comment (jkotas) [inline src/coreclr/src/vm/genericdict.cpp]: This should not be needed with the simpler scheme.

File: src/coreclr/src/vm/genericdict.cpp
Comment (jkotas) [inline src/coreclr/src/vm/genericdict.cpp]: Why not to update derived types lazily as well, only once they need the dictionary slot?

File: src/coreclr/src/inc/corinfo.h
Comment (BruceForstall) [reply src/coreclr/src/inc/corinfo.h]: @fadimounir How close is this to getting merged? Have you updated the JIT-EE interface GUID?

Comment (fadimounir) [reply src/coreclr/src/inc/corinfo.h]: @BruceForstall I have a couple of code review feedback workitems that I still need to work on. It'll probably be a day or two. Were you trying to merge other PRs and waiting on this one?

Comment (BruceForstall) [reply src/coreclr/src/inc/corinfo.h]: Looks like https://github.com/dotnet/runtime/pull/31686 got closed without merging, so I guess is abandoned. This makes this the last known JIT-EE interface change in flight.

I was just hoping to batch all the JIT-EE interface changes quickly. But, of course, take the time to get it right.

File: src/coreclr/src/vm/generics.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/generics.cpp]: Fixed

File: src/coreclr/src/vm/genericdict.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/genericdict.cpp]: It might be possible that thread A allocates a new dictionary, copies the entries from the old dictionary, and update the m_pPerInstInfo pointer to that new dictionary, and thread B sees the new pointer, and attempts to read slots from that new dictionary, but gets NULL/invalid values from the cpu pipeline because the memcpy hasn't fully flushed the changes. Or am I misunderstanding something?

File: src/coreclr/src/vm/genericdict.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/genericdict.cpp]: As discussed, I will look into it and make the updates on the derived types lazy as well.

File: src/coreclr/src/vm/genericdict.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/genericdict.cpp]: Shouldn't be really...

Comment (fadimounir) [reply src/coreclr/src/vm/genericdict.cpp]: I'll remove it

File: src/coreclr/src/vm/genmeth.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/genmeth.cpp]: Ok, I'll move these helpers to class Dictionary

File: src/coreclr/src/vm/class.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/class.cpp]: As discussed, I change this to be lazily updated as well whenever we call `JIT_GenericHandleWorker`

File: src/coreclr/src/vm/genericdict.cpp
Comment (jkotas) [reply src/coreclr/src/vm/genericdict.cpp]: > gets NULL/invalid values from the cpu pipeline 

That cannot happen because of [data dependency ordering](https://www.bing.com/search?q=Data+Dependency+Ordering). The read of the dictionary slot depends on the read of the pointer to the dictionary. The data dependency ordering is enforced all hardware that we care about. We depend on it in many places.

Comment (jkotas) [reply src/coreclr/src/vm/genericdict.cpp]: Note that `FlushProcessWriteBuffers` is pretty expensive operation. It sends interrupt to all cores. You want to use it only when you really have to.

File: src/coreclr/src/vm/genericdict.cpp
Comment (jkotas) [inline src/coreclr/src/vm/genericdict.cpp]: Loader allocator returns zeroed out memory. This is not needed.

File: src/coreclr/src/vm/genericdict.cpp
Comment (jkotas) [inline src/coreclr/src/vm/genericdict.cpp]: This may be growing the dictionaries unnecessarily. Should this check whether the dictionary that we got is big enough for slotIndex first, and grow the dictionary only when it is not?

File: src/coreclr/src/vm/genericdict.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/genericdict.cpp]: Ok thanks for the clarification Jan and confirming that the invalid state can no longer happen. I will remove this flushing operation.

File: src/coreclr/src/vm/genericdict.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/genericdict.cpp]: Thanks Jan, yes that's a good point. We can defer expansion until we hit a slot that is actually beyond the size of the dictionary. I'll make that update.

File: src/coreclr/src/vm/genericdict.cpp
Comment (jkotas) [inline src/coreclr/src/vm/genericdict.cpp]: Do you need to take this lock even when we are not resizing the dictionary (typical case)?

File: src/coreclr/src/vm/genericdict.cpp
Comment (jkotas) [inline src/coreclr/src/vm/genericdict.cpp]: This will turn linear algorithm into quadratic algorithm for debug builds. Should we limit the amount of stress we do in debug builds? E.g. by growing by 1 only up to 10 slots, and the switch to doubling?

File: src/coreclr/src/vm/jitinterface.cpp
Comment (jkotas) [inline src/coreclr/src/vm/jitinterface.cpp]: Where did this magic number `1` (and `2` below) came from? I know it was there even before your change, but I am not able to figure it out.

File: src/coreclr/src/vm/method.hpp
Comment (jkotas) [inline src/coreclr/src/vm/method.hpp]: Nit: Reword why this works without saying safe vs. unsafe

File: src/coreclr/src/vm/methodtablebuilder.cpp
Comment (jkotas) [inline src/coreclr/src/vm/methodtablebuilder.cpp]: Do we ever have non-null dictionary layout with MaxSlots == 0?

File: src/coreclr/src/vm/genericdict.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/genericdict.cpp]: No. I missed that. We can move the lock to the point where we actually resize

File: src/coreclr/src/vm/genericdict.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/genericdict.cpp]: We could. I've had the +1 suggested in the original PR for debug builds as a way to stress the feature :)

File: src/coreclr/src/vm/jitinterface.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/jitinterface.cpp]: For methods: we indirect once of the MethodDesc to get the m_pPerInstInfo (1)
For types: we indirect twice of the MethodTable to get the dictionary. m_pPerInstInfo is an array of dictionaries for types, due to inheritance properties.

File: src/coreclr/src/vm/methodtablebuilder.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/methodtablebuilder.cpp]: I remember hit such cases before when working on the original PR, but can't remember specifics at this point. I'll have to debug again to remember

File: src/coreclr/src/vm/generics.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/generics.cpp]: Yes. We should always be dealing with non-zero slots here. I'll remove the check

File: src/coreclr/src/vm/jitinterface.cpp
Comment (jkotas) [reply src/coreclr/src/vm/jitinterface.cpp]: Ah ok ... make sense.

File: src/coreclr/src/vm/generics.cpp
Comment (fadimounir) [reply src/coreclr/src/vm/generics.cpp]: Made it into an assert.

File: src/coreclr/src/vm/genericdict.cpp
Comment (jkotas) [inline src/coreclr/src/vm/genericdict.cpp]: Nit: This can be regular store. No other threads has access to the new entries at this point.

**Review by jkotas — APPROVED** [review]
Thanks!

**Pull request merged**
